STEP02_Step1 과제 리팩토링

gangfunction opened the pull request:
실수로 풀 리퀘스트를 처음에 커밋을 잘못해서 이렇게 나온것같아요. 
전반적인 전체 개발과정은 리드미에 똑같이 적어놨으니 혹시 리뷰하실때 필요하면 참조해주시면 감사하겠습니다... ㅎㅎ
현재 가진생각은 커밋을 의미단위로 쪼개는게 좀 힘든것 같네요.  트래킹이 잘못되있는것도 히스토리 날려도 어느정도는 남아있는거 같아서 그렇게 생각했습니다.  

제가 가진 질문은 DTO객체에서 생성시 값이 유효한지 체크를 하는데요. 이걸 서비스 구현체에서 이중으로 검증할 필요가 있을까요? 아니면 확실하게 검증하기위해서 각 영역마다 어떻게 가져가는게 유연하면서 확장성이 좋을까요? 

그리고 제가 가진 코드스타일에 대한 느낌적인 평가를 해주시면 감사하겠습니다.. 지금 최대한 가독성이 좋게 그리고 책임분리를 명확하게 할려고 하는데,  오히려 둘다 못챙긴 것같기도하고... 생각이 많아지네요 ㅋㅋ

gangfunction committed:
updated

Took 2 minutes

gangfunction committed:
for review

gangfunction committed:
updated

Took 2 minutes

gangfunction committed:
updated

gangfunction committed:
DTO created

UserPointDTO
PointHistoryDTO

Took 24 minutes

gangfunction committed:
dto변환 테스트 추가

Took 31 minutes

gangfunction committed:
서비스단에서 Optional 객체 처리

Took 33 minutes

gangfunction committed:
타임아웃 로직 추가후 타임아웃에 대한 에러메시지가 출력되는 테스트 추가

Took 1 hour 23 minutes

gangfunction committed:
리패키징과 중복로직 리팩토링

Took 3 hours 43 minutes

gangfunction committed:
DTO 로직 개선

dto를 레코드 클래스를 활용하여 개선했습니다.

Took 24 minutes

gangfunction committed:
updated

gangfunction committed:
@Scheduler에서 ThreadTaskScheduler로 변경

1. queue size 와 queue processing 실행 딜레이 혼합 사용
2. System 클래스의 시간 사용보다 -> java.Time클래스의 Instant.now() 를 활용
3. 테스트 코드 각 클래스에 맞게 이관 작업
4. 통합테스트 구축 (문제 살짝 있음)

gangfunction committed:
주석 추가및 레이어화

gangfunction committed:
Repository 영역 테스트 추가 및 README 업데이트

gangfunction committed:
gitignore updated

gangfunction committed:
Update .gitignore

gangfunction committed:
git ignore updated

hubtwork commented on src/main/java/io/hhplus/tdd/point/repository/PointRepository.java:
```diff
@@ -0,0 +1,41 @@
+   * @param id : 사용자 아이디
+   * @param amount : 포인트
+   */
+  void insertOrUpdate(long id, long amount);
+
+  /**
+   * 포인트 히스토리를 삽입
+   * @param id : 사용자 아이디
+   * @param amount : 포인트
+   * @param type : 트랜잭션 타입
+   * @param updateMillis : 업데이트 시간
+   */
+  void insertHistory(long id, long amount, TransactionType type, long updateMillis);
+
```
삽입 후 해당 데이터를 반환하여 유저에게 명시적으로 알려줄 수 있는 구조를 만드는게 좋지 않을까요 ?

gangfunction replied:
삽입후 반환하는 구조로 변경했습니다.!
  private UserPoint updateUserPointAndHistory(UserPoint userPoint, long currentAmount,
      TransactionType charge) {
    UserPoint outputPoint = pointRepository.insertOrUpdate(userPoint.id(), currentAmount);
    PointHistory pointHistory = pointRepository.insertHistory(userPoint.id(), userPoint.point(),
        charge,
        Instant.now().toEpochMilli());
    if(pointHistory == null){
      log.error("History not found");
      throw new IllegalArgumentException("History not found");
    }
    return outputPoint;
  }
  다만 히스토리 삽입은 단건 조회가 없기에  있는지 없는지만 확인하도록 했습니다..!

hubtwork commented on src/main/java/io/hhplus/tdd/point/repository/PointRepository.java:
```diff
@@ -0,0 +1,41 @@
+   * @param id : 사용자 아이디
+   * @return Optional<List<PointHistory>> : 포인트 히스토리 도메인 객체 리스트
+   */
+  Optional<List<PointHistory>> selectHistories(long id);
```
`select` 라는 워딩보다 가져온다는 개념의 `get` 이라는 워딩이 좀 더 명확한 것 같아요. ( `select` 는 SQL 의 Language 인데, Repository 명세의 대상이 DB 일지 아닐지 모르니까 )

gangfunction replied:
말씀대로 변경했습니다..! 
  @Override
  public Optional<List<PointHistory>> getHistories(long id) {
    return Optional.of(pointHistoryTable.selectAllByUserId(id));
  }
  말그대로 repository의 명세의 대상이 불분명한거같아 이해했습니다.

hubtwork commented on src/main/java/io/hhplus/tdd/point/repository/PointRepositoryImpl.java:
```diff
@@ -0,0 +1,62 @@
+@Repository
+public class PointRepositoryImpl implements PointRepository{
+  PointHistoryTable pointHistoryTable = new PointHistoryTable();
+  UserPointTable userPointTable = new UserPointTable();
```
각 Table 은 `@Component` 를 통해 Bean 으로 올라가 관리될 텐데, 왜 여기서는 new 로 직접 할당했을까요 ? Spring 이 관리하고 있는 Bean 이랑은 다른 객체가 되어버리겠네요.

hubtwork commented on src/main/java/io/hhplus/tdd/point/service/QueueManager.java:
```diff
@@ -0,0 +1,191 @@
+  private final ChargeSpecification chargeImpl;
+  private final UseSpecification useImpl;
+
+  // * ThreadTaskScheduler를 사용하여 비동기 처리를 위한 스케줄러를 생성합니다.
```
Queue 를 통한 접근을 하신 분은 처음 봐서 새롭네요 👍 
그런데 QueueManager 는 특정 Task 에 의존적이지 않아야하는 거 같은데, 직접적으로 Repository , Charge 등 비즈니스 로직을 주입받아 활용하고 있습니다. 이는 좋은 방법이 아닌 거 같아요.

hubtwork replied:
각 비즈니스 로직에 비의존적으로 활용가능하도록 하고, 사용하는 측면에서 queueing 할 수 있도록 제공되면 좋을 것 같아요.

hubtwork commented on src/main/java/io/hhplus/tdd/point/service/history/HistorySpecification.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.List;
+
+public interface HistorySpecification {
```
불필요한 추상화가 너무 많은 것 같아요. 이 경우, 의사 Depth 가 깊어져서 가독성을 오히려 저해시킬 수 있어요.

gangfunction replied:
![스크린샷 2024-06-24 14 16 34](https://github.com/gangfunction/hhplus-tdd-java/assets/62240333/8b3bca8d-f130-4336-b02c-707f64f67b0c)
말씀대로 불필요한 추상화들을 제거했습니다..! 하지만 이럴때 패키징을 어떻게 가져가야할지 아직 고민중입니다.

hubtwork commented on src/test/java/io/hhplus/tdd/point/controller/PointControllerTest.java:
```diff
@@ -0,0 +1,121 @@
+            patch("/point/{id}/charge", userId)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(objectMapper.writeValueAsString(userPoint)))
+        .andExpect(status().isOk())
+        .andExpect(jsonPath("$.id").value(userId))
+        .andExpect(jsonPath("$.point").value(amount));
+    verify(pointService).charge(userPoint);
```
jsonPath 로 항목별로 검증할 경우, 누락 건이 있더라도 다른 값이 일치하다면 테스트를 성공하는 불상사가 생길 수 있어요. 이 부분은 jsonString 으로 변환해 일치여부를 검증하도록 테스트를 구성하는 게 좋은 것 같습니다.

gangfunction replied:
![스크린샷 2024-06-24 14 17 48](https://github.com/gangfunction/hhplus-tdd-java/assets/62240333/339ede9d-b4c2-463c-9670-3e20ace0c734)
마찬가지로 객체로 변환해 검증하는걸로 변경했고, String 구조를 명시해둬서 테스트를 추가할 예정입니다.!

hubtwork commented on src/test/java/io/hhplus/tdd/point/service/QueueConcurrencyServiceTest.java:
```diff
@@ -0,0 +1,226 @@
+ * 4. 논 리피터블 리드 : 같은 쿼리를 반복할때 업데이트 쿼리로 인하여 결과가 달라지는 문제
+ * 5. 팬텀 읽기 : 같은 쿼리를 반복했으나 예상되지않은 값을 읽어오는 문제
+ */
+public class QueueConcurrencyServiceTest {
```
`@SpringBootTest` 로 실제 Spring 의 의존성 주입을 받아 테스트 하는게 더 바람직해보입니다.

hubtwork commented:
책임 분리 등은 열심히 한 것 같은데, 너무 원자적으로 분리해놓아 코드의 추적이 어렵고 가독성이 떨어집니다. 또한 파편화가 많이 이루어져 있는 코드의 경우, 일부 변화가 너무 큰 범위에 변경 영향을 끼칠 수 있는데, 이를 놓칠 수 있어요.

gangfunction commented:
리뷰 감사합니다. 과도한 원자화로 응집도를 좀 깨뜨린것같다는 생각을 해보고 수정해야겠네요 ㅎㅎ

gangfunction committed:
동기화 로직 취약점 발견 및, 개선중

풀리퀘스트에 따른 수정사항은 수정 완료했습니다. 다만 처리를함에있어서 사용자가 포인트를 조회한 시점에서 그 전의 문제들이 처리가 되지않는 문제를 발견했습니다.

Took 6 hours 35 minutes
